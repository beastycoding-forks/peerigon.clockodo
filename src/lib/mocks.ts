import { faker } from "@faker-js/faker";

export const ONE_DAY = 24 * 60 * 60 * 1000;
export const ONE_YEAR = 356 * ONE_DAY;

const MAX_ITERATION_COUNT = 10000;

export const isoDateFromDateTime = (dateTime: Date) => {
  return [
    dateTime.getFullYear(),
    String(dateTime.getMonth() + 1).padStart(2, "0"),
    String(dateTime.getDate()).padStart(2, "0"),
  ].join("-");
};

export const isoMonthFromDateTime = (dateTime: Date) => {
  return [
    dateTime.getFullYear(),
    String(dateTime.getMonth() + 1).padStart(2, "0"),
  ].join("-");
};

export const isoDateFromTimestamp = (timestamp: number) => {
  return isoDateFromDateTime(new Date(timestamp));
};

export const timestampFromIsoDate = (isoDate: string) => {
  return new Date(isoDate).getTime();
};

export const endOfMonth = (date: Date) => {
  return new Date(date.getFullYear(), date.getMonth() + 1, 0, 23, 59, 59, 999);
};

export const endOfYear = (dateTime: Date) => {
  return new Date(dateTime.getFullYear() + 1, 0, 0, 23, 59, 59, 999);
};

export const nextDay = (date: Date) => {
  const nextDay = new Date(date.getTime() + ONE_DAY);

  return new Date(nextDay.getFullYear(), nextDay.getMonth(), nextDay.getDate());
};

export const generateUnique = <Value>({
  count,
  generate,
}: {
  count: number;
  generate: () => Value;
}) => {
  const generated = new Set<Value>();

  for (let i = 0; generated.size < count; i++) {
    if (i === MAX_ITERATION_COUNT) {
      throw new Error(
        `Couldn't generate enough unique values before reaching the max iteration count.
This usually happens when the generated values create too many conflicts (e.g. too many dates on a small date range).`
      );
    }
    generated.add(generate());
  }

  return generated;
};

export const generateRandomDates = ({
  count,
  between: [from, to],
}: {
  count: number;
  between: [Date, Date];
}) => {
  return Array.from(
    generateUnique({
      count,
      generate: () => {
        const randomDate = faker.date.between(from, to);

        return new Date(
          randomDate.getFullYear(),
          randomDate.getMonth(),
          randomDate.getDate()
        ).getTime();
      },
    })
  ).sort();
};

export const generateRandomMonths = ({
  count,
  between: [from, to],
}: {
  count: number;
  between: [Date, Date];
}) => {
  return Array.from(
    generateUnique({
      count,
      generate: () => {
        const randomDate = faker.date.between(from, to);

        return new Date(
          randomDate.getFullYear(),
          randomDate.getMonth()
        ).getTime();
      },
    })
  ).sort();
};

export const toPairs = <Item>(array: Array<Item>) => {
  if (array.length % 2 !== 0) {
    throw new Error("Cannot create pairs: Array length must be divisible by 2");
  }

  const count = array.length / 2;
  const pairs = new Array<[Item, Item]>(count);

  for (let i = 0; i < count; i++) {
    pairs[i] = [array[i * 2], array[i * 2 + 1]];
  }

  return pairs;
};
