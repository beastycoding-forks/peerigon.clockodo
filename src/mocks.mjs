// eslint-disable-next-line import/no-unresolved
export * from "./mocks.js";
