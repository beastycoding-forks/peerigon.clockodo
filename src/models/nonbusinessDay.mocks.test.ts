import { createNonbusinessDaysMocks } from "./nonbusinessDay.mocks.js";

describe("createNonbusinessDaysMocks()", () => {
  test("It creates realistic mocks", () => {
    expect(createNonbusinessDaysMocks({ count: 20 })).toMatchSnapshot();
  });

  test("It re-uses ids for the next year", () => {
    const [nonbusinessDay1, nonbusinessDay2] = createNonbusinessDaysMocks({
      count: 2,
      dateBetween: [new Date(2020, 11, 31), new Date(2021, 0, 2)],
    });

    expect(nonbusinessDay1.id === nonbusinessDay2.id);
  });
});
