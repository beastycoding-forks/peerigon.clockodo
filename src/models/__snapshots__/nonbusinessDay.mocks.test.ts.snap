// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createNonbusinessDaysMocks() It creates realistic mocks 1`] = `
Array [
  Object {
    "date": "2020-02-21",
    "halfDay": false,
    "id": 0,
    "name": "et qui quia",
  },
  Object {
    "date": "2020-03-24",
    "halfDay": false,
    "id": 1,
    "name": "et inventore eligendi",
  },
  Object {
    "date": "2020-04-14",
    "halfDay": false,
    "id": 2,
    "name": "deserunt id harum",
  },
  Object {
    "date": "2020-05-05",
    "halfDay": false,
    "id": 3,
    "name": "odio quia vitae",
  },
  Object {
    "date": "2020-05-23",
    "halfDay": false,
    "id": 4,
    "name": "quo provident molestiae",
  },
  Object {
    "date": "2020-05-26",
    "halfDay": false,
    "id": 5,
    "name": "ullam pariatur quos",
  },
  Object {
    "date": "2020-05-30",
    "halfDay": false,
    "id": 6,
    "name": "quod laborum ea",
  },
  Object {
    "date": "2020-06-03",
    "halfDay": false,
    "id": 7,
    "name": "esse quia expedita",
  },
  Object {
    "date": "2020-06-05",
    "halfDay": false,
    "id": 8,
    "name": "autem tempora ut",
  },
  Object {
    "date": "2020-06-25",
    "halfDay": false,
    "id": 9,
    "name": "doloribus harum voluptas",
  },
  Object {
    "date": "2020-06-28",
    "halfDay": false,
    "id": 10,
    "name": "omnis et accusantium",
  },
  Object {
    "date": "2020-07-20",
    "halfDay": false,
    "id": 11,
    "name": "quia quos sequi",
  },
  Object {
    "date": "2020-07-31",
    "halfDay": false,
    "id": 12,
    "name": "beatae enim necessitatibus",
  },
  Object {
    "date": "2020-09-07",
    "halfDay": true,
    "id": 13,
    "name": "placeat saepe eligendi",
  },
  Object {
    "date": "2020-09-09",
    "halfDay": false,
    "id": 14,
    "name": "excepturi et laudantium",
  },
  Object {
    "date": "2020-09-11",
    "halfDay": false,
    "id": 15,
    "name": "similique fugit corporis",
  },
  Object {
    "date": "2020-09-17",
    "halfDay": true,
    "id": 16,
    "name": "dolores esse nobis",
  },
  Object {
    "date": "2020-09-20",
    "halfDay": false,
    "id": 17,
    "name": "sint magnam aut",
  },
  Object {
    "date": "2020-10-12",
    "halfDay": false,
    "id": 18,
    "name": "dolores maiores incidunt",
  },
  Object {
    "date": "2020-12-24",
    "halfDay": false,
    "id": 19,
    "name": "iusto laborum facere",
  },
]
`;
